ocaml/typing/jkind.ml

```diff
@@ -261,6 +274,238 @@ end
+
+  let void = Sort Sort.void
+
+  let float64 = Sort Sort.float64
+
+  module Debug_printers = struct
+    open Format
+
+    let t ppf = function
+      | Any -> fprintf ppf "Any"
+      | Sort s -> fprintf ppf "Sort %a" Sort.Debug_printers.t s
+  end
+end
+
+(* Whether or not a type is external to the garbage collector *)
+module Externality = struct
+  type t =
+    | External (* not managed by the garbage collector *)
+    | External64 (* not managed by the garbage collector on 64-bit systems *)
+    | Internal (* managed by the garbage collector *)
+
+  (* CR layouts v2.8: Either use this or remove it *)
+  let _to_string = function
+    | External -> "external"
+    | External64 -> "external64"
+    | Internal -> "internal"
+
+  let max = Internal
+
+  let min = External
+
+  let equal e1 e2 =
+    match e1, e2 with
+    | External, External -> true
+    | External64, External64 -> true
+    | Internal, Internal -> true
+    | (External | External64 | Internal), _ -> false
+
+  let sub t1 t2 : Sub_result.t =
+    match t1, t2 with
+    | External, External -> Equal
+    | External, (External64 | Internal) -> Sub
+    | External64, External -> Not_sub
+    | External64, External64 -> Equal
+    | External64, Internal -> Sub
+    | Internal, (External | External64) -> Not_sub
+    | Internal, Internal -> Equal
+
+  let intersection t1 t2 =
+    match t1, t2 with
+    | External, (External | External64 | Internal)
+    | (External64 | Internal), External ->
+      External
+    | External64, (External64 | Internal) | Internal, External64 -> External64
+    | Internal, Internal -> Internal
+
+  module Debug_printers = struct
+    open Format
+
+    let t ppf = function
+      | External -> fprintf ppf "External"
+      | External64 -> fprintf ppf "External64"
+      | Internal -> fprintf ppf "Internal"
+  end
+end
+
+module Const = struct
+  type t =
+    { layout : Layout.Const.t;
+      externality : Externality.t
+    }
+
+  (* CR layouts v2.8: remove this *)
+  let to_legacy_jkind { layout; externality } : Jane_asttypes.const_jkind =
+    match layout, externality with
+    | Any, _ -> Any
+    | Sort Value, Internal -> Value
+    | Sort Value, External64 -> Immediate64
+    | Sort Value, External -> Immediate
+    | Sort Void, _ -> Void
+    | Sort Float64, _ -> Float64
+
+  (* CR layouts v2.8: do a better job here *)
+  let to_string t = const_jkind_to_string (to_legacy_jkind t)
+
+  let sub { layout = lay1; externality = ext1 }
+      { layout = lay2; externality = ext2 } =
+    Sub_result.combine (Layout.Const.sub lay1 lay2) (Externality.sub ext1 ext2)
+end
+
+module Desc = struct
+  type t =
+    | Const of Const.t
+    | Var of Sort.var (* all modes will be [max] *)
+
+  let format ppf =
+    let open Format in
+    function
+    | Const c -> fprintf ppf "%s" (Const.to_string c)
+    | Var v -> fprintf ppf "%s" (Sort.var_name v)
+
+  (* considers sort variables < Any, but otherwise just checks for equality.
```

Comment:
I think this comment is slightly out of date wrt externality, which makes it no longer just an equality check.

Reply:
I've clarified the comment. Better?